Implement `Tabs` in editor settings

chad1008 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
Related: #52997

## What?
<!-- In a few words, what is the PR actually doing? -->
Implement the new `Tabs` component in the editor settings sidebar

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Due to its more granular nature, `Tabs` is better suited to more applications than its predecessor, `TabPanel`. Implementing it here will ensure it gets all of the accessibility and other benefits, and ongoing support, of the packaged component. This move also cuts down on the amount of custom CSS needed, as we're using a component with its own styles already built in that no longer need to be duplicated.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
This PR modifies `Tabs` slightly by exposing the context that the component uses to pass data to its subcomponents. The editor settings sidebar implements an internal `ComplimentaryArea`, which in turn has an internal `Slot/Fill` that we need to pass our context through.

Now that we are exposing `Tabs.Context` we can use the relevant provider to pass the context wherever it needs to go.

## Testing Instructions
1. Open the post editor
2. Ensure the editor settings sidebar on the right hand side of the screen opens and closes as expected
3. Test both tabs (Post and Block). Ensure they display the correct sidebar/controls
4. In the main editor panel, click between a block and the post title.
5. Ensure that when you click on a block, the Block sidebar is automatically selected. When you click the post title, the Post sidebar should be displayed.
6. Open a new page in the editor. Select whatever template you'd like (we're going to edit it in a minute).
7. Confirm the tabs and editor-based selection you tested on a post also work the same way here.
8. On the Page sidebar, click on the name of the currently selected template, then click "Edit Template" in the modal that appears.
9. Confirm that the sidebar correctly enters "Template mode", showing you a Template tab instead of a Page tab.
10. Confirm once again that clicking on a block activates the Block sidebar, selecting that tab automatically.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

1. Open a new post in the post editor.
2. Enter text into the title field (should be focused by default)
3. Press Enter/Return to move to the first paragraph block
4. Confirm that the sidebar updated to the Block tab when the block was focused
5. Press Up Arrow to return to the Title field
6. Confirm that the Post tab is automatically selected
7. Press Tab to focus the sidebar
8. Use the arrow keys to navigate between buttons and Enter/Return/Space to activate a button and view that tab

### Known Issues

- While writing the above testing steps, I discovered the `Tabs` implementation is breaking the close button on the sidebar. Activating that button via mouse or keyboard triggers a `maximum update depth` error from React. That button is provided by the `ComplementaryAreaHeader` subcomponent. I'm looking into why and how to address that.
- By default, Tabs should place the focus on the `tablist` element, and then navigate between individual `tabs` with arrow keys. Arrowing to a tab should activate it automatically. With this implementation, when tabbing over to the sidebar the focus lands directly on the first `tab` (which semantically is a `button`). You then use the arrow keys to move between tabs/buttons, but because the actual button is focused and not the `tablist`, you need to explicitly activate the button with enter/return/space. I suspect this is being caused Tabs being nested inside a `ComplementaryArea` but need to look into it further.

chad1008 committed:
implement Tabs in editor settings sidebar

chad1008 committed:
remove duplicated styles from editor sidebar

ciampo commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -27,66 +33,68 @@ const SettingsHeader = ( { sidebarName } ) => {
 			: [ documentLabel, '' ];
 
-	const [ blockAriaLabel, blockActiveClass ] =
+	const [ blockAriaLabel ] =
 		sidebarName === 'edit-post/block'
 			? // translators: ARIA label for the Block Settings Sidebar tab, selected.
 			  [ __( 'Block (selected)' ), 'is-active' ]
 			: // translators: ARIA label for the Block Settings Sidebar tab, not selected.
 			  [ __( 'Block' ), '' ];
```
Since we're not destructuring the active class anymore, we can also remove it from the lines below

```suggestion
	const blockAriaLabel =
		sidebarName === 'edit-post/block'
			? // translators: ARIA label for the Block Settings Sidebar tab, selected.
			  __( 'Block (selected)' )
			: // translators: ARIA label for the Block Settings Sidebar tab, not selected.
			  __( 'Block' );
```

Same for the `templateAriaLabel` below

ciampo commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -27,66 +33,68 @@ const SettingsHeader = ( { sidebarName } ) => {
+					id={ 'edit-post/document' }
+					render={
+						isTemplateMode ? (
+							<Button
```
Is it necessary to `render` those `Tabs` as `Button`s?

chad1008 replied:
Good point. It is at the moment, but that's only because `Tabs.Tab` (and the other subcomponents) doesn't currently accept and forward all of the props being applied here. That's something we'd [already talked about improving](https://github.com/WordPress/gutenberg/pull/53960#pullrequestreview-1661817575) but I haven't done you. I'll get a separate PR together for that so this `render` prop is no longer necessary 👍

chad1008 replied:
PR to incorporate all HTML element props into the subcomponents: #55860

I've tested locally and it eliminates the need for these `render` props, so once that's merged, I can rebase and update this PR.

chad1008 replied:
With that PR merged, and this one rebased, I've removed the `render`s so the props can pass down normally ✅

ciampo commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -27,66 +33,68 @@ const SettingsHeader = ( { sidebarName } ) => {
+						isTemplateMode ? (
+							<Button
+								onClick={ openDocumentSettings }
+								className={ `edit-post-sidebar__panel-tab` }
```
Is this classname still necessary, now that the tab styles come directly from the `Tabs` component?

chad1008 replied:
Good point, I don't think it is. The only place this classname is in use that I can see is [part of a style override in a different sidebar.](https://github.com/WordPress/gutenberg/blob/12c17e509ce4d5bce3f7e811f283520499a24880/packages/edit-post/src/components/secondary-sidebar/style.scss#L36).

ciampo commented on packages/edit-post/src/components/sidebar/settings-sidebar/index.js:
```diff
@@ -69,36 +73,54 @@ const SettingsSidebar = () => {
 	);
 
+	const Content = () => {
+		const tabsContextValue = useContext( Tabs.Context );
```
Let's add a comment about why we're forwarding the tabs context here

ciampo reviewed (commented):
> * While writing the above testing steps, I discovered the `Tabs` implementation is breaking the close button on the sidebar. Activating that button via mouse or keyboard triggers a `maximum update depth` error from React. That button is provided by the `ComplementaryAreaHeader` subcomponent. I'm looking into why and how to address that.

Interesting. Sounds like the click handler there is triggering some sort of infinite loop.

> * By default, Tabs should place the focus on the `tablist` element, and then navigate between individual `tabs` with arrow keys. Arrowing to a tab should activate it automatically. With this implementation, when tabbing over to the sidebar the focus lands directly on the first `tab` (which semantically is a `button`).

When looking at the [Tabs storybook example](https://wordpress.github.io/gutenberg/iframe.html?args=&id=components-experimental-tabs--default&viewMode=story), I can't see the tablist element receiving focus. Pressing tab seems to move focus directly to the selected tab.

> You then use the arrow keys to move between tabs/buttons, but because the actual button is focused and not the `tablist`, you need to explicitly activate the button with enter/return/space. I suspect this is being caused Tabs being nested inside a `ComplementaryArea` but need to look into it further.

I'm not sure I understand exactly what is going on. My expectation is that pressing arrow keys when the individual `tab` buttons are focused should move the tab selection (no need for `tablist` to be focused).

chad1008 committed:
incorporate initial feedback

chad1008 commented:
> When looking at the [Tabs storybook example](https://wordpress.github.io/gutenberg/iframe.html?args=&id=components-experimental-tabs--default&viewMode=story), I can't see the tablist element receiving focus. Pressing tab seems to move focus directly to the selected tab.

Yeah, you're right, sorry. I was misremembering. Focus does go to the buttons. Here's what the difference in keyboard navigation looks like, after looking more closely:

`trunk` (legacy behavior):
- The `tab` key moves focus to the first tab/button.
- To move between tabs you use `tab`/`shift+tab`.
- It **is** possible to focus the currently selected tab by pressing the `tab` key.
- Once focused via keyboard action, tabs must be explicitly activated with `return/enter/space`
- Focus doesn't wrap around the `tablist` when tabbing past the first/last tab.

`Tabs` in Storybook (expected behavior):
- The `tab` key moves focus to the first tab/button.
- To move between tabs you use `left arrow/right arrow`.
- It **is** possible to focus the currently selected tab by pressing the `arrow` key to get back to it.
- Once focused via keyboard action, tabs are automatically activated with (unless `selectOnMove` is set to `false`)
- Focus does wrap around the `tablist` when arrowing past the first/last tab

this PR:
- ✅ The `tab` key moves focus to the first tab/button.
- ✅ To move between tabs you use `left arrow/right arrow`.
- ⚠️ It **is not** possible to focus the currently selected tab by pressing the `arrow` key to get back to it. The active tab is just skipped over.
- ⚠️ Once focused via keyboard action, tabs must be explicitly activated with `return/enter/space`
- ⚠️ Focus does **not** wrap around the `tablist` when arrowing past the first/last tab

So the focus is indeed going to the right place, but navigation and tab selection for keyboard users are a bit off when compared to expected `Tabs` behavior.

ciampo commented:
I wonder if that's related to `Slot`/`Fill` and how events are bubbling. To test my hypothesis, it would be interesting to refactor the code and experiment adding `bubblesVirtually` to the `Slot` (not sure if any other changes would be needed off the top of my head).
Another counterproof that this is related to `Slot`/`Fill` could be to temporarily remove those components (without caring too much about visual UI changes), and see if the component reverts back to behaving as expected.

chad1008 committed:
pass props to Tabs directly

chad1008 commented:
> Another counterproof that this is related to `Slot`/`Fill` could be to temporarily remove those components (without caring too much about visual UI changes), and see if the component reverts back to behaving as expected.

I know when you and I looked at this together the other day, `bubblesVirtually` didn't seem to solve the problem, but I thought the idea of removing the `Slot`/`Fill` from the equation was an interesting one. To test it, I removed the `ComplementaryAreaFill` sub-component entirely, so the `Tabs` components could just render normally.

<details>
<summary>My test diff:</summary>

```diff
diff --git a/packages/interface/src/components/complementary-area/index.js b/packages/interface/src/components/complementary-area/index.js
index 887c447d92..aa82f0e298 100644
--- a/packages/interface/src/components/complementary-area/index.js
+++ b/packages/interface/src/components/complementary-area/index.js
@@ -196,14 +196,7 @@ function ComplementaryArea( {
 				</ComplementaryAreaMoreMenuItem>
 			) }
 			{ isActive && (
-				<ComplementaryAreaFill
-					className={ classnames(
-						'interface-complementary-area',
-						className
-					) }
-					scope={ scope }
-					id={ identifier.replace( '/', ':' ) }
-				>
+				<div className={ 'interface-complementary-area' }>
 					<ComplementaryAreaHeader
 						className={ headerClassName }
 						closeLabel={ closeLabel }
@@ -244,7 +237,7 @@ function ComplementaryArea( {
 						) }
 					</ComplementaryAreaHeader>
 					<Panel className={ panelClassName }>{ children }</Panel>
-				</ComplementaryAreaFill>
+				</div>
 			) }
 		</>
 	);

```

</details>

It messes up the UI placement, and the tab sequence is a bit awkward this way... but once things aren't being routed through a `Fill`, but here's what I found:

- Still cannot arrow key to the currently selected tab. It gets skipped. (consistent with this PRs current issues)
- Tabs are not automatically activated like they should be. (consistent with this PRs current issues)
- Manually activating a tab automatically moves the focus to the next tab. (new, and not expected behavior)
- Closing the sidebar is still broken.

 So removing the `Fill` feels like it left keyboard navigation even _more_ broken, which I'll admit surprises me.

I was talking this over with @brookewp yesterday as well, and she made an interesting observation. Keyboard navigation on this PR is behaving a little like the `Tabs` component has `selectOnMove` set `false`, aka "manual activation." If you look at the [Manual Activation story in storybook](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-tabs--manual-activation), the `data-focus-visible` and `data-active-item` attributes both flow to whatever button is focused as you arrow between tabs. The tab doesn't get _selected_, but it does register as both focused and active. On this PR, as you arrow between tabs `data-focus-visible` behaves this way, but `data-active-item` does not. That attribute doesn't apply until the tab is manually selected, which is another departure from expected `Tabs` behavior.

@ciampo, you shared the concern with me that it's possible some of Ariakit's internal store/context is what isn't making it through the `ComplementaryArea` internal `Slot`/`Fill` and causing these unexpected behaviors. I'm not sure if the data attribute discovery Brooke made points us more in that direction or not, but it's all very odd at this point.

@jsnajdr, I know you're pretty well-versed in `Slot`/`Fill` issues, if you have a chance to look any of this over, does anything jump out at you that I might want to look into? Happy to chat/investigate directly if that would help outline the problem space more clearly.

_Edit to add:_ I did try to recreate the problem in a simplified example as well, albeit unsuccessfully. There's a temporary [Slot Fill Problem Example](http://localhost:50240/?path=/story/components-experimental-tabs--slot-fill-problem-example) pushed to this PR that tries to replicate the issue, even on that story, complete with `slot`/`fill` I wasn't able to recreate the keyboard behavior issues.

jsnajdr commented:
> the concern with me that it's possible some of Ariakit's internal store/context is what isn't making it through the ComplementaryArea internal Slot/Fill and causing these unexpected behaviors.

The first thing to check is what type of slot/fill your ComplementaryArea is. Is it the "bubblesVirtually" type, or the "classic" type?

With "bubblesVirtually", a component is rendered in the React context of the Fill, although in DOM it's rendered under the Slot's elements. It doesn't see the Slot's React context. So, if the Slot is rendered as a child of some Ariakit component, the parent's Ariakit context is not visible to it.

This is a common problem, and we've discussed this with @ciampo before, mainly in #51264 which introduced a new forwaredContext API.

chad1008 commented:
> The first thing to check is what type of slot/fill your ComplementaryArea is. Is it the "bubblesVirtually" type, or the "classic" type?

Thanks @jsnajdr! This particular `slot`/`fill` is not the `bubbleVirtually` variety, though @ciampo and I did experiment with that while pairing, and it didn't seem to resolve the issue.

ciampo commented on packages/edit-post/src/components/sidebar/settings-sidebar/index.js:
```diff
@@ -69,36 +73,57 @@ const SettingsSidebar = () => {
+				isActiveByDefault={ SIDEBAR_ACTIVE_BY_DEFAULT }
+			>
+				<Tabs.Context.Provider value={ tabsContextValue }>
+					<Tabs.TabPanel id={ 'edit-post/document' }>
```
Should the `TabPanel`s have `focusable={false}`?

chad1008 replied:
Good call, thank you!

ciampo commented on packages/edit-post/src/components/sidebar/settings-sidebar/index.js:
```diff
@@ -69,36 +73,57 @@ const SettingsSidebar = () => {
+				isActiveByDefault={ SIDEBAR_ACTIVE_BY_DEFAULT }
+			>
+				<Tabs.Context.Provider value={ tabsContextValue }>
+					<Tabs.TabPanel id={ 'edit-post/document' }>
```
It would be good to refactor these sidebar IDs to a variable and reuse them across the code (instead of having hardcoded strings). For example, in this file alone, those IDs are used both in the `TabPanel` components, but also inside the `useSelect` select hooks a few lines above. They are also used in the `SettingsHeader` component and assigned as ids to the `Tab` components.

chad1008 replied:
I looked around a bit to see if these sidebar names are already abstracted somewhere they could be imported from, but didn't come up with anything, so I went with refactoring them into a variable myself.

<details>
<summary>Is along the lines of what you were thinking?</summary>

```diff
diff --git a/packages/edit-post/src/components/sidebar/settings-header/index.js b/packages/edit-post/src/components/sidebar/settings-header/index.js
index 2633b55cec..bf1a39fb89 100644
--- a/packages/edit-post/src/components/sidebar/settings-header/index.js
+++ b/packages/edit-post/src/components/sidebar/settings-header/index.js
@@ -11,6 +11,7 @@ import { store as editorStore } from '@wordpress/editor';
  */
 import { store as editPostStore } from '../../../store';
 import { unlock } from '../../../lock-unlock';
+import { sidebars } from '../settings-sidebar';
 
 const { Tabs } = unlock( componentsPrivateApis );
 
@@ -26,20 +27,20 @@ const SettingsHeader = ( { sidebarName } ) => {
 	}, [] );
 
 	const documentAriaLabel =
-		sidebarName === 'edit-post/document'
+		sidebarName === sidebars.document
 			? // translators: ARIA label for the Document sidebar tab, selected. %s: Document label.
 			  sprintf( __( '%s (selected)' ), documentLabel )
 			: documentLabel;
 
 	const blockAriaLabel =
-		sidebarName === 'edit-post/block'
+		sidebarName === sidebars.block
 			? // translators: ARIA label for the Block Settings Sidebar tab, selected.
 			  __( 'Block (selected)' )
 			: // translators: ARIA label for the Block Settings Sidebar tab, not selected.
 			  __( 'Block' );
 
 	const templateAriaLabel =
-		sidebarName === 'edit-post/document'
+		sidebarName === sidebars.document
 			? __( 'Template (selected)' )
 			: __( 'Template' );
 
@@ -47,7 +48,7 @@ const SettingsHeader = ( { sidebarName } ) => {
 		<>
 			<Tabs.TabList>
 				<Tabs.Tab
-					id={ 'edit-post/document' }
+					id={ sidebars.document }
 					aria-label={
 						isTemplateMode ? templateAriaLabel : documentAriaLabel
 					}
@@ -58,7 +59,7 @@ const SettingsHeader = ( { sidebarName } ) => {
 					{ isTemplateMode ? __( 'Template' ) : documentLabel }
 				</Tabs.Tab>
 				<Tabs.Tab
-					id={ 'edit-post/block' }
+					id={ sidebars.block }
 					aria-label={ blockAriaLabel }
 					// translators: Data label for the Block Settings Sidebar tab.
 					data-label={ __( 'Block' ) }
diff --git a/packages/edit-post/src/components/sidebar/settings-sidebar/index.js b/packages/edit-post/src/components/sidebar/settings-sidebar/index.js
index b662fb79ee..630348e4d3 100644
--- a/packages/edit-post/src/components/sidebar/settings-sidebar/index.js
+++ b/packages/edit-post/src/components/sidebar/settings-sidebar/index.js
@@ -37,6 +37,10 @@ const SIDEBAR_ACTIVE_BY_DEFAULT = Platform.select( {
 	web: true,
 	native: false,
 } );
+export const sidebars = {
+	document: 'edit-post/document',
+	block: 'edit-post/block',
+};
 
 const SettingsSidebar = () => {
 	const { sidebarName, isSidebarOpen, keyboardShortcut, isTemplateMode } =
@@ -52,15 +56,11 @@ const SettingsSidebar = () => {
 				editPostStore.name
 			);
 			const isOpen = sidebar !== null;
-			if (
-				! [ 'edit-post/document', 'edit-post/block' ].includes(
-					sidebar
-				)
-			) {
+			if ( ! [ sidebars.document, sidebars.block ].includes( sidebar ) ) {
 				if ( select( blockEditorStore ).getBlockSelectionStart() ) {
-					sidebar = 'edit-post/block';
+					sidebar = sidebars.block;
 				}
-				sidebar = 'edit-post/document';
+				sidebar = sidebars.document;
 			}
 			const shortcut = select(
 				keyboardShortcutsStore
@@ -96,10 +96,7 @@ const SettingsSidebar = () => {
 				isActiveByDefault={ SIDEBAR_ACTIVE_BY_DEFAULT }
 			>
 				<Tabs.Context.Provider value={ tabsContextValue }>
-					<Tabs.TabPanel
-						id={ 'edit-post/document' }
-						focusable={ false }
-					>
+					<Tabs.TabPanel id={ sidebars.document } focusable={ false }>
 						{ ! isTemplateMode && (
 							<>
 								<PostStatus />
@@ -115,7 +112,7 @@ const SettingsSidebar = () => {
 						) }
 						{ isTemplateMode && <TemplateSummary /> }
 					</Tabs.TabPanel>
-					<Tabs.TabPanel id={ 'edit-post/block' } focusable={ false }>
+					<Tabs.TabPanel id={ sidebars.block } focusable={ false }>
 						<BlockInspector />
 					</Tabs.TabPanel>
 				</Tabs.Context.Provider>

```

</details>

Of course, if we end up removing `aria-label` and/or `data-label`, some of this would get removed. I just wanted to confirm this is what you had in mind.

ciampo replied:
Thinking about this, it's probably better to leave it:

- there are definitely more possible values for the active complementary area than just these two
- I wonder if those values should be defined in a more "central" place, maybe even in the interface store

Those changes would be out of scope for this PR, and therefore it's probably best to keep those hardcoded values as they currently are (which was also the case before this PR).

ciampo commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -27,66 +30,51 @@ const SettingsHeader = ( { sidebarName } ) => {
+				<Tabs.Tab
+					id={ 'edit-post/document' }
+					onClick={ openDocumentSettings }
+					aria-label={
+						isTemplateMode ? templateAriaLabel : documentAriaLabel
+					}
+					data-label={
+						isTemplateMode ? __( 'Template' ) : documentLabel
+					}
```
Now that we have proper tabs semantics, it would be great to delete those extra attributes on tabs:
- We should understand why this `data-label` attribute is necessary. I suspect it could be because of tests, in which case we could simply rewrite tests by querying by role and reading its accessible label
- It looks like the purpose of `aria-label` was to add "(Selected)" when the tab is selected for assistive technology users. That shouldn't be necessary anymore, since the the `Tabs` component should already take care of that

chad1008 replied:
I took a quick look into tests and didn't immediately see the `data-label` in use there, but I didn't deep dive. I went back a while in the history and that attribute has been there for a long time now, but I didn't dig back to its initial implementation. I'll look a bit more to try and determine if/how it's in use, then if it's just tests, update them accordingly.

> It looks like the purpose of aria-label was to add "(Selected)" when the tab is selected for assistive technology users. That shouldn't be necessary anymore, since the the Tabs component should already take care of that

Do you specifically mean appending "(Selected)" to the `aria-label`? Or just that we'll have a valid `aria-selected` attribute applied to the current tab, so assistive tech is appropriately notified?

ciampo replied:
> I'll look a bit more to try and determine if/how it's in use, then if it's just tests, update them accordingly.

Sounds good. Alternatively, we could also just delete it and see if tests fail because of it.

> Do you specifically mean appending "(Selected)" to the `aria-label`? Or just that we'll have a valid `aria-selected` attribute applied to the current tab, so assistive tech is appropriately notified?

The latter. What I meant is that passing a custom `aria-label` attribute shouldn't be necessary anymore, since we're now using `Tabs` (which takes care of this internally by setting the `aria-selected="true"` attribute as needed).

chad1008 replied:
Perfect, I wanted to make sure I was on the same page regarding the `aria-label`!

I tracked down when `data-label` was [first implemented back in 2018](https://github.com/WordPress/gutenberg/pull/9793). It was being used so it could populate the `content` of a psuedoelement that was used for styling that has since been removed, so we should be safe to remove it.

I've just pushed an update that does away with both of these attributes. Still working well locally.

ciampo commented on packages/components/src/tabs/index.tsx:
Changes to `Tabs` (including the memoization of the context value) should probably happen in a separate PR

chad1008 replied:
Agreed. I had a couple other very minor clean-up tasks for the component, and I've [combined them with the changes we need for this PR](https://github.com/WordPress/gutenberg/pull/56224).

ciampo commented on packages/interface/src/lock-unlock.js:
We should be able to remove lock-unlock changes related to the `@wordpress/interface` package

ciampo reviewed (commented):
I took some time to look into this properly, and found out what is causing the issues highlighted above (spoiler alert — it's not `Slot`/`Fill`!)

### Problem 1: keyboard interaction on tabs is broken

When I started looking at the symptoms I was particularly curious about the fact that, while keyboard interactions were not working, clicking individual tabs was working.

I then noticed that the `Tab` components have dedicated `onClick` handlers, which made me realize that the `Tabs` component is not being controlled correctly. The right way to control the component is via the `onSelect` callback on the `Tabs` component, while there's no need for those `onClick` handlers.

<details>
<summary>I tested those changes on my machine, and all interactions on the tabs started working again (click to expand)</summary>

```diff
diff --git a/packages/edit-post/src/components/sidebar/settings-header/index.js b/packages/edit-post/src/components/sidebar/settings-header/index.js
index 37045cb697..6c5644acdb 100644
--- a/packages/edit-post/src/components/sidebar/settings-header/index.js
+++ b/packages/edit-post/src/components/sidebar/settings-header/index.js
@@ -15,11 +15,6 @@ import { unlock } from '../../../lock-unlock';
 const { Tabs } = unlock( componentsPrivateApis );
 
 const SettingsHeader = ( { sidebarName } ) => {
-	const { openGeneralSidebar } = useDispatch( editPostStore );
-	const openDocumentSettings = () =>
-		openGeneralSidebar( 'edit-post/document' );
-	const openBlockSettings = () => openGeneralSidebar( 'edit-post/block' );
-
 	const { documentLabel, isTemplateMode } = useSelect( ( select ) => {
 		const postTypeLabel = select( editorStore ).getPostTypeLabel();
 
@@ -53,7 +48,6 @@ const SettingsHeader = ( { sidebarName } ) => {
 			<Tabs.TabList>
 				<Tabs.Tab
 					id={ 'edit-post/document' }
-					onClick={ openDocumentSettings }
 					aria-label={
 						isTemplateMode ? templateAriaLabel : documentAriaLabel
 					}
@@ -65,7 +59,6 @@ const SettingsHeader = ( { sidebarName } ) => {
 				</Tabs.Tab>
 				<Tabs.Tab
 					id={ 'edit-post/block' }
-					onClick={ openBlockSettings }
 					aria-label={ blockAriaLabel }
 					// translators: Data label for the Block Settings Sidebar tab.
 					data-label={ __( 'Block' ) }
diff --git a/packages/edit-post/src/components/sidebar/settings-sidebar/index.js b/packages/edit-post/src/components/sidebar/settings-sidebar/index.js
index e8681205b6..be654060a6 100644
--- a/packages/edit-post/src/components/sidebar/settings-sidebar/index.js
+++ b/packages/edit-post/src/components/sidebar/settings-sidebar/index.js
@@ -5,8 +5,8 @@ import {
 	BlockInspector,
 	store as blockEditorStore,
 } from '@wordpress/block-editor';
-import { useSelect } from '@wordpress/data';
-import { Platform, useContext } from '@wordpress/element';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { Platform, useContext, useCallback } from '@wordpress/element';
 import { isRTL, __ } from '@wordpress/i18n';
 import { drawerLeft, drawerRight } from '@wordpress/icons';
 import { store as interfaceStore } from '@wordpress/interface';
@@ -73,6 +73,8 @@ const SettingsSidebar = () => {
 		[]
 	);
 
+	const { openGeneralSidebar } = useDispatch( editPostStore );
+
 	const Content = () => {
 		// Because `PluginSidebarEditPost` renders a `ComplementaryArea`, we
 		// need to forward the `Tabs` context so it can be passed through the
@@ -120,8 +122,17 @@ const SettingsSidebar = () => {
 		);
 	};
 
+	const onTabsSelect = useCallback(
+		( newSelectedTabId ) => {
+			if ( !! newSelectedTabId ) {
+				openGeneralSidebar( newSelectedTabId );
+			}
+		},
+		[ openGeneralSidebar ]
+	);
+
 	return (
-		<Tabs selectedTabId={ sidebarName }>
+		<Tabs selectedTabId={ sidebarName } onSelect={ onTabsSelect }>
 			<Content />
 		</Tabs>
 	);

```
</details>

### Problem 2: closing the sidebar causes the editor to crash

When looking at this issue, I started looking at the console error, which suggested that the component rendering infinitely was the complementary area's `Fill`, which is where the tabs are rendered.

Looking at the changes that this PR introduces, one of the first suspects for causing those extra re-renders was the part, in the `SettingsSidebar` component, where we have to manually forward the `Tabs.Context` to the `SettingsHeader` component. I tried to comment the context provider out, and closing the sidebar started working again.

Now, that couldn't be the fix, but at least it was a big hint: the issue seems to be related to the `Tabs` component and its context. I then took a look at the `Tabs` component, and immediately noticed that [the internal context value is not memorized](https://github.com/WordPress/gutenberg/blob/274fe0050badf47f8d10dc881c0255303621ae2e/packages/components/src/tabs/index.tsx#L158), which means that a new object is going to be created every time the component re-renders. I wrapped the context value in a `useMemo` hook, and reloaded the editor.

Closing the sidebar still crashed, but the error message was different. Instead of happening in the `Fill` component, not the infinite rendering loop was happening directly in the `Tabs` component. I therefore started looking at the code inside the `Tabs` component, in particular at the three `useLayoutEffect` hooks. [One of them, in particular, runs only when the component is used in controlled mode](https://github.com/WordPress/gutenberg/blob/274fe0050badf47f8d10dc881c0255303621ae2e/packages/components/src/tabs/index.tsx#L138-L155). I commented it out, and closing the sidebar started to work again!

So, what is happening? Basically, when the sidebar is closed, all the different tabs are removed from the page, which means that [this line of code can't find a selected tab](https://github.com/WordPress/gutenberg/blob/274fe0050badf47f8d10dc881c0255303621ae2e/packages/components/src/tabs/index.tsx#L58) (since the array of `items` is empty). And therefore, [this line of code runs](https://github.com/WordPress/gutenberg/blob/274fe0050badf47f8d10dc881c0255303621ae2e/packages/components/src/tabs/index.tsx#L147), trying to reset the selected tab id to `null`. Depending on how the controlled `Tabs` implementation was coded (see how I solved problem 1), this may cause an infinite loop. Therefore, the solution that I found here was to pass `selectedTabId={null}` to `Tabs` when the sidebar is closed.

<details>
<summary>These are changes that I applied on my machine (click to expand)</summary>

```diff
diff --git a/packages/components/src/tabs/index.tsx b/packages/components/src/tabs/index.tsx
index 826c4e7c9d..5273e4c21f 100644
--- a/packages/components/src/tabs/index.tsx
+++ b/packages/components/src/tabs/index.tsx
@@ -8,7 +8,7 @@ import * as Ariakit from '@ariakit/react';
  * WordPress dependencies
  */
 import { useInstanceId } from '@wordpress/compose';
-import { useLayoutEffect, useRef } from '@wordpress/element';
+import { useLayoutEffect, useMemo, useRef } from '@wordpress/element';
 
 /**
  * Internal dependencies
@@ -154,8 +154,13 @@ function Tabs( {
 		setSelectedId,
 	] );
 
+	const contextValue = useMemo(
+		() => ( { store, instanceId } ),
+		[ store, instanceId ]
+	);
+
 	return (
-		<TabsContext.Provider value={ { store, instanceId } }>
+		<TabsContext.Provider value={ contextValue }>
 			{ children }
 		</TabsContext.Provider>
 	);
diff --git a/packages/edit-post/src/components/sidebar/settings-sidebar/index.js b/packages/edit-post/src/components/sidebar/settings-sidebar/index.js
index be654060a6..89a756d258 100644
--- a/packages/edit-post/src/components/sidebar/settings-sidebar/index.js
+++ b/packages/edit-post/src/components/sidebar/settings-sidebar/index.js
@@ -39,8 +39,8 @@ const SIDEBAR_ACTIVE_BY_DEFAULT = Platform.select( {
 } );
 
 const SettingsSidebar = () => {
-	const { sidebarName, keyboardShortcut, isTemplateMode } = useSelect(
-		( select ) => {
+	const { sidebarName, isSidebarOpen, keyboardShortcut, isTemplateMode } =
+		useSelect( ( select ) => {
 			// The settings sidebar is used by the edit-post/document and edit-post/block sidebars.
 			// sidebarName represents the sidebar that is active or that should be active when the SettingsSidebar toggle button is pressed.
 			// If one of the two sidebars is active the component will contain the content of that sidebar.
@@ -51,6 +51,7 @@ const SettingsSidebar = () => {
 			let sidebar = select( interfaceStore ).getActiveComplementaryArea(
 				editPostStore.name
 			);
+			const isOpen = sidebar !== null;
 			if (
 				! [ 'edit-post/document', 'edit-post/block' ].includes(
 					sidebar
@@ -66,12 +67,11 @@ const SettingsSidebar = () => {
 			).getShortcutRepresentation( 'core/edit-post/toggle-sidebar' );
 			return {
 				sidebarName: sidebar,
+				isSidebarOpen: isOpen,
 				keyboardShortcut: shortcut,
 				isTemplateMode: select( editPostStore ).isEditingTemplate(),
 			};
-		},
-		[]
-	);
+		}, [] );
 
 	const { openGeneralSidebar } = useDispatch( editPostStore );
 
@@ -132,7 +132,10 @@ const SettingsSidebar = () => {
 	);
 
 	return (
-		<Tabs selectedTabId={ sidebarName } onSelect={ onTabsSelect }>
+		<Tabs
+			selectedTabId={ isSidebarOpen ? sidebarName : null }
+			onSelect={ onTabsSelect }
+		>
 			<Content />
 		</Tabs>
 	);

```
</details>

chad1008 committed:
add `closeGeneralSidebar` to `onSelect` callback

chad1008 committed:
set TabPanels to `focuasable={false}`

chad1008 committed:
detect when sidebar is closed. pass `selectedTabId` of `null`

chad1008 committed:
improve `Tabs` `onSelect` callback

chad1008 commented:
Thank you for debugging there @ciampo! I've updated the `onSelect` callback from what you and I looked at when we last spoke to match the more-complete version I see in your first diff.

I was following a similar thread for the closed-sidebar-loop-of-infinite-doom, but you arrived at a speedier and more effective solution than I did, so I've applied those changes as well.

I believe I've applied or replied to all open comments, but please let me know if I've missed anything.

Oh, there are some things that will no longer work with this PR. Specifically, anything that relied on changes to `Tabs` itself, like the exposure of `Tabs.Context` or memoizing that context value. Once #56224 merges I can rebase this PR.

chad1008 commented:
rebased ✅

ciampo commented on packages/edit-post/src/components/sidebar/settings-sidebar/index.js:
```diff
@@ -62,43 +67,80 @@ const SettingsSidebar = () => {
-			identifier={ sidebarName }
-			header={ <SettingsHeader sidebarName={ sidebarName } /> }
-			closeLabel={ __( 'Close Settings' ) }
-			headerClassName="edit-post-sidebar__panel-tabs"
```
We should search for the classname around the codebase and:

- for e2e tests matches, change that to ideally use more semantic selectors (if needed, we could look into adding an accessible label to the tablist)
- for style matches, delete the styles that are not needed anymore after the reactor to Tabs, and keep the ones that are still useful
- make sure that there aren't any instances of tabs using this same classname (if there are, re-assess the above with this new info in mind)

chad1008 replied:
This comment started with "Regardless" which make me worry I missed a step and I want to be sure I'm following you fully. Regardless of what?

ciampo replied:
Nevermind that "Regardless", it's a remnant of a previous version of that comment that I forgot to clean up 😅

chad1008 replied:
Pushed updates related to tests and styles using this classname, including a thought/comment on my approach to one of the changes.

ciampo commented on packages/edit-post/src/components/sidebar/settings-sidebar/index.js:
```diff
@@ -62,43 +67,80 @@ const SettingsSidebar = () => {
-			icon={ isRTL() ? drawerLeft : drawerRight }
-			isActiveByDefault={ SIDEBAR_ACTIVE_BY_DEFAULT }
+		<Tabs
+			selectedTabId={ isSidebarOpen ? sidebarName : null }
```
We should probably leave a comment to why we're passing `null` when the sidebar isn't open, to spare the same headache to the next developer touching this file

chad1008 replied:
done!

ciampo reviewed (commented):
Tabs seem to be working well!

I that there's a double border now showing:

![Screenshot 2023-11-17 at 20 01 01](https://github.com/WordPress/gutenberg/assets/1083581/2fc311cb-779c-444e-bace-f15daddb91c0)

That's because we're now wrapping each tab panel contents into a `TabPanel` component (where previously we had a react fragment, ie. no dom element). The extra `div` rendered by the `TabPanel` between the `Panel` and the `PanelBody` causes [this rule](https://github.com/WordPress/gutenberg/blob/3afffb3070afe17437842120670d819d140e9067/packages/components/src/panel/style.scss#L6-L8) not to apply anymore, which causes the top border of the first panel body to show.

The simplest solution (without involving style overrides) IMO could be to pass a `className` to `PluginSidebarEditPost`, through which we apply a `margin-top: -1px` style 🤞  We should add a comment to why that is necessary, though.


Apart from that, I think that this PR can be marked as ready for review. It would be also be great if @alexstine could take a look (I know he's been waiting for this part of the editor to have accessible tabs in a while!)

chad1008 committed:
remove `aria-label` and `data-label` props

chad1008 committed:
add note explaining null selected tab when sidebar is closed

chad1008 commented:
Marking as ready for review, but wanted to note that there will be some tests to update before this merges.

chad1008 commented:
Started looking at some of the e2e tests that need attention and found a new wrinkle: on smaller viewports, the sidebar toggle doesn't work. The sidebar will be closed by default on mobile devices, but opening it isn't possible with this PR applied.

There aren't any errors to chase, but after if I slowly implement changes from this PR, it looks like wrapping `PluginSidebarEditPost` in an internal `Content` component is what's triggering the breakage. I'll need to investigate more next week.

ciampo commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -27,66 +26,18 @@ const SettingsHeader = ( { sidebarName } ) => {
-				</Button>
-			</li>
-		</ul>
+		<>
```
Nit: It looks like this fragment isn't necessary

alexstine reviewed (commented):
@chad1008 Giving this an A11Y sign off with screen readers. I will allow others to review for technical feedback.

youknowriad commented:
I'd like to understand the story of the "Tabs" component. Do we plan to deprecate "TabsPanel" (or reimplement internally using "Tabs"). What's the plan here?

chad1008 commented:
> I'd like to understand the story of the "Tabs" component. Do we plan to deprecate "TabsPanel" (or reimplement internally using "Tabs"). What's the plan here?

There's some additional context/discussion in #52997, which I just noticed I hadn't linked to in this PRs description, so I've just updated that, thanks!

The broad plan for the component is to migrate `TabPanel` implementations, and faux tab-patterns like this one, to `Tabs` across the editor. This one was actually intentionally chosen as our fist migration for its complexity. We wanted to use it to stress-test the new components and reveal places to improve them. Having done that, and added some improvements, we'll be tackling other implementations alongside wrapping up this PR.

Once we've converted all of the editor's `TabPanel` implementations over to `Tabs`, the plan is to deprecate `TabPanel`.

chad1008 committed:
update e2e test

chad1008 committed:
style updates

chad1008 committed:
update internal component structure to avoid rerenders

chad1008 commented on packages/e2e-tests/specs/editor/various/preferences.test.js:
```diff
@@ -46,7 +50,7 @@ describe( 'preferences', () => {
 		// Dismiss.
 		await page.click(
-			'.edit-post-sidebar__panel-tabs [aria-label="Close Settings"]'
+			'div[aria-label="Editor settings"] div[role="tablist"] + button[aria-label="Close Settings"]'
```
I don't love relying on a sibling selector here. There are technically two `button[aria-label="Close Settings"]`, and being adjacent to the `tablist` is what makes the one we want stand out the most without falling back to a classname. I'm not actually sure this is better, it feels fragile. 🤔

ciampo replied:
The fact that there are two "Close Settings" button should have also been the case before the changes from this PR. It looks like removing the `display: none` rule in `packages/edit-post/src/components/sidebar/style.scss` could be responsible for this test regression.

If re-instatig the `display: none` styles solve the "there are two buttons with the same label" situation, we could either keep the selector unchanged from `trunk` (since the `edit-post-sidebar__panel-tabs` is still being used), or at most change it to

```suggestion
			'div[aria-label="Editor settings"] button[aria-label="Close Settings"]'
```

chad1008 replied:
I don't think reinstating that `display: none` would help here. I actually only removed that attribute from the styles because it was [already being declared elsewhere](https://github.com/WordPress/gutenberg/blob/341473bb0949893aeb8c658cc255898745bb7425/packages/interface/src/components/complementary-area-header/style.scss#L22). So the button was always `display:none`, with or without that specific change.

You are right, there have always been two, of these buttons. The previous `.edit-post-sidebar__panel-tabs` selector only targeted one. If targeting by classname is preferable here I'm happy to fall back to that. The only reason for this change was wanting to target aria labels and/or roles when possible. 😄

ciampo replied:
Makes sense! Then I guess we can keep the selector that you wrote. Is the test still failing, though?

chad1008 replied:
Nope, this test should be passing just fine

chad1008 committed:
remove fragment

chad1008 commented:
Update: Looking at some of the test failures, one issue we're hitting is a conflict with sidebars that can be registered by third parties like plugins.

[This test](https://github.com/WordPress/gutenberg/blob/369947cc16a1b8803bc6499d1e1a3fbd0f20915f/test/e2e/specs/editor/plugins/annotations.spec.js), for example, [registers](https://github.com/WordPress/gutenberg/blob/8066e00777a761e2916363e8f9e30c27a562aaf8/packages/e2e-tests/plugins/plugins-api/annotations-sidebar.js) such a sidebar. When the test attempts to click the button to open that sidebar, the editor crashes due to a maximum-depth error.

After debugging, the part of this PR causing the issue looks to be putting `Tabs` into controlled mode. Once `Tabs` is given a valid `selectedTabId`, the underlying `PluginSidebarEditPost` component is unable to load any other sidebars. I'll need to play with it a bit more to come up with a fix.

I have some holiday stuff going on between now and monday, but I have no doubt my curiosity will get the better of me, so I'll be checking into this a little when I have some downtime.

ciampo commented on packages/edit-post/src/components/sidebar/style.scss:
```diff
@@ -1,20 +1,8 @@
-	}
 
 	.components-button.has-icon {
-		display: none;
```
Related to https://github.com/WordPress/gutenberg/pull/55360/files#r1403026209

ciampo commented on packages/edit-post/src/components/sidebar/style.scss:
```diff
@@ -1,20 +1,8 @@
-	}
 
 	.components-button.has-icon {
-		display: none;
-		margin: 0 0 0 auto;
```
Related to https://github.com/WordPress/gutenberg/pull/55360/files#r1403026209

chad1008 committed:
prevent infinite loop when opening third party sidebar

chad1008 commented:
I did take a look over the weekend, and found a solution to the third-party/plugin sidebar problem. I've pushed those changes, waiting to see what that does for the CI checks 🤞

chad1008 committed:
update e2e tests for `Tabs` compatibility

chad1008 commented:
There were some remaining failures related to implementation details and/or how `Tabs` handles tab stops differently than the standalone buttons of the previous implementation, and I've updated those tests accordingly.

Still failing: 
 
`editor/various/keyboard-navigable-blocks.spec.js › Order of block keyboard navigation › allows tabbing in navigation mode if no block is selected`

To replicate the failure yourself:
1. Create a new post. Add two paragraph blocks with a little text in each (the words `one` and `two` are fine)
2. Click an empty space on the editor canvas to remove focus from any specific block
3. Press `Tab` three times. This should cycle focus through the title, the first paragraph, and land on the second paragraph
4. Press `Tab` one more time and observe where the focus lands in the settings sidebar

- previous/expected behavior: focus lands on the the `Post` tab, which is currently selected, because when the sidebar was focused there was no block selected
- current behavior: while the `Post` tab is properly selected because there's no block selected, and its contents are displayed. But focus goes to the `Block` tab.

I think this is due to a timing issue. `Tabs` (specifically Ariakit under the hood) applies a `tabindex` of `-1` to all tabs besides the currently active one. When you tab through the final paragraph block, the `Block` tab is active because a block is selected. When you hit `Tab`, the focus goes to the next focusable element, which is the `Block` tab, (`Post` has a `-1` `tabindex` at this point. As soon as focus shifts off of the block, the sidebar's listener's kick in and change the selected tab to `Post`, despite that not being where the focus is going.

Previously, both `buttons` were focusable via `tab`, but now only the active button is. Moving tabs is done with arrow keys. 

The end result is that someone relying on keyboard nav and/or assisted tech ends up in a state where the focused tab doesn't match the content actually being displayed. That doesn't feel ideal to me, but I'm curious what others think. I can look into ways to ensure focus follows the currently selected tab under the right combinations of controlled mode and `selectOnMove`, unless this feels low priority from an a11y perspective. cc @alexstine / @andrewhayward

chad1008 commented:
@diegohaz I could use your input here, just to confirm that what I'm seeing is expected behavior within `Ariakit.Tab`

To summarize the above issue:

1. `Tabs` is being implemented in the settings sidebar, using `Ariakit.Tab` internally, in controlled mode.
2. Focus is on a block in the Gutenberg editor, which means the **Block** tab in our sidebar is selected automatically due to some internal logic.
3. Pressing the [Tab] key moves focus from the Gutenberg block to the currently selected tab,(**Block**) as I'd expect.
4. As this happens, some of Gutenberg's internal logic updates the selected tab from the **Block** tab to the **Post** tab, because no block is currently selected once the focus moves.
5. Focus does not follow the selected tab change, and stays on the no-longer-selected **Block** tab.
6. The net effect for the user is that focus lands on a tab that does not reflect the content actually being displayed, which I worry might be confusing for utilizing assistive tech.

I can simplify and replicate this in Storybook, by adding a delay to the Controlled Mode story's dropdown actions. If I trigger a delayed change to the `selectedId`, and then focus the current tab with the [Tab] key, focus does not follow the selection change when it happens. That makes sense to me, focus probably shouldn't bounce around unexpectedly, but in our current implementation we might want to control it a bit more.

<details>
<summary>If it's helpful, this here's a diff to replicate my test in Storybook:</summary>

1. Apply this against `trunk` and navigate to the `Tabs` Controlled Mode story
2. Use the dropdown to select a new tab
3. During the three second delay, click on the canvas and hit [Tab] to send focus to the currently selected tab
4. Watch the values logged to the console as the selected tab is changed out from underneath your focus

Now, if I monitor (in Storybook) the `activeId`, the appears to update along with the delayed change to `selectedId`, even though the actual focus does _not_ update. I was considering trying to use `setActiveId()` to "pin" the focus to the currently selected tab in specific controlled-mode cases, but that doesn't seem to be possible.

```diff
diff --git a/packages/components/src/tabs/index.tsx b/packages/components/src/tabs/index.tsx
index 12dd0b4fcc..ede66a51ee 100644
--- a/packages/components/src/tabs/index.tsx
+++ b/packages/components/src/tabs/index.tsx
@@ -44,7 +44,7 @@ function Tabs( {
 
 	const isControlled = selectedTabId !== undefined;
 
-	const { items, selectedId } = store.useState();
+	const { items, selectedId, activeId } = store.useState();
 	const { setSelectedId } = store;
 
 	// Keep track of whether tabs have been populated. This is used to prevent
@@ -154,6 +154,13 @@ function Tabs( {
 		setSelectedId,
 	] );
 
+	useLayoutEffect( () => {
+		if ( ! isControlled ) {
+			return;
+		}
+		console.log( { selectedId, activeId } );
+	}, [ isControlled, selectedId, activeId ] );
+
 	const contextValue = useMemo(
 		() => ( {
 			store,
diff --git a/packages/components/src/tabs/stories/index.story.tsx b/packages/components/src/tabs/stories/index.story.tsx
index ce8c8324ed..2a724c0171 100644
--- a/packages/components/src/tabs/stories/index.story.tsx
+++ b/packages/components/src/tabs/stories/index.story.tsx
@@ -273,17 +273,29 @@ const ControlledModeTemplate: StoryFn< typeof Tabs > = ( props ) => {
 					<DropdownMenu
 						controls={ [
 							{
-								onClick: () => setSelectedTabId( 'tab1' ),
+								onClick: () =>
+									setTimeout(
+										() => setSelectedTabId( 'tab1' ),
+										3000
+									),
 								title: 'Tab 1',
 								isActive: selectedTabId === 'tab1',
 							},
 							{
-								onClick: () => setSelectedTabId( 'tab2' ),
+								onClick: () =>
+									setTimeout(
+										() => setSelectedTabId( 'tab2' ),
+										3000
+									),
 								title: 'Tab 2',
 								isActive: selectedTabId === 'tab2',
 							},
 							{
-								onClick: () => setSelectedTabId( 'tab3' ),
+								onClick: () =>
+									setTimeout(
+										() => setSelectedTabId( 'tab3' ),
+										3000
+									),
 								title: 'Tab 3',
 								isActive: selectedTabId === 'tab3',
 							},

```

</details>

Does that sound like what you'd expect? I'm mostly confused by `activeId` and the currently focused element getting out of sync? It's possible I'm misunderstanding what the [Ariakit docs](https://ariakit.org/reference/tab-provider#activeid) mean when they say say `activeId` references the id of the currently focused item.

diegohaz commented:
@chad1008 Indeed, [`setSelectedId`](https://ariakit.org/reference/use-tab-store#setselectedid) isn't intended to move focus. It simply affects the `selectedId` state, which is then translated into `aria-selected` and other state attributes. Even [`setActiveId`](https://ariakit.org/reference/use-tab-store#setactiveid-1) on its own doesn't impact DOM focus. It updates the state and some state attributes like [`data-active-item`](https://ariakit.org/guide/styling#data-active-item). We use it, for instance, when hovering over menu items and other composite items with the [`focusOnHover`](https://ariakit.org/reference/composite-hover#focusonhover) prop.

The method that moves DOM focus in composite stores is [`move`](https://ariakit.org/reference/use-tab-store#move). It updates the `activeId` state and moves DOM focus. For tabs, we also have the [`select`](https://ariakit.org/reference/use-tab-store#select) method, which is essentially `move` combined with `setSelectedId`.

This is already mentioned in the docs for the methods. We should probably provide further clarification in the `activeId` documentation as well.

chad1008 committed:
fix double top margin on tabpanels

chad1008 commented:
> Even [`setActiveId`](https://ariakit.org/reference/use-tab-store#setactiveid-1) on its own doesn't impact DOM focus.
> The method that moves DOM focus in composite stores is [`move`](https://ariakit.org/reference/use-tab-store#move). It updates the `activeId` state and moves DOM focus. For tabs, we also have the [`select`](https://ariakit.org/reference/use-tab-store#select) method, which is essentially `move` combined with `setSelectedId`.

Awesome. I was experimenting with `setActiveId` when I noticed the behavior I described, now I understand why it was throwing me off. Thank you! I'll look into `move` and/or `select`. Sorry I didn't spot those before pinging you!

Thanks again!

chad1008 commented:
Okay, potential solution to the focus vs selection race condition described above in #56658

Once that, or some other solution, is merged, we should be pretty close to rebasing thing thing to completion!

chad1008 committed:
update to use new tabId prop

chad1008 committed:
remove import that is no longer needed after rebase

chad1008 committed:
fix keyboard navigable blocks test

chad1008 committed:
fix visibility and tab order tests

chad1008 committed:
fix footnotes tests

ciampo reviewed (commented):
Some e2e tests are still failing, could it be because the test selector is looking for the `edit-post-sidebar__panel-tab` classname, which doesn't seem to be used anymore in this PR?

chad1008 committed:
fix change detection test

chad1008 commented:
> Some e2e tests are still failing, could it be because the test selector is looking for the edit-post-sidebar__panel-tab classname, which doesn't seem to be used anymore in this PR?

Yep! I actually just popped back on after dinner to wrap up (what *should* be) the last test failure. That now-defunct classname was the culprit in some other spots as well. Checks are running now 🤞 

This PR has also been updated and rebased to account for the new controlled mode focus behavior, and the new `tabId` prop.

chad1008 commented:
Green is my new favorite color.

ciampo reviewed (approved):
🚀
